File: src/java/test/src/androidTest/java/com/tencent/wcdbtest/winq/WindowDefTest.java
// Created by chenqiuwen on 2023/6/10.
//

/*
 * Tencent is pleased to support the open source community by making
 * WCDB available.
 *
 * Copyright (C) 2017 THL A29 Limited, a Tencent company.
 * All rights reserved.
 *
 * Licensed under the BSD 3-Clause License (the "License"); you may not use
 * this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 *       https://opensource.org/licenses/BSD-3-Clause
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.tencent.wcdbtest.winq;

import androidx.test.ext.junit.runners.AndroidJUnit4;
import static com.tencent.wcdbtest.base.WinqTool.winqEqual;

import com.tencent.wcdb.winq.Column;
import com.tencent.wcdb.winq.Expression;
import com.tencent.wcdb.winq.ExpressionConvertible;
import com.tencent.wcdb.winq.FrameSpec;
import com.tencent.wcdb.winq.Order;
import com.tencent.wcdb.winq.WindowDef;

import org.junit.Test;
import org.junit.runner.RunWith;

@RunWith(AndroidJUnit4.class)
public class WindowDefTest {
    @Test
    public void test() {
        winqEqual(new WindowDef(), "");
        winqEqual(new WindowDef().partitionBy("column1"),
                "(PARTITION BY column1)");
        winqEqual(new WindowDef().partitionBy(new String[]{"column1", "column2"}),
                "(PARTITION BY column1, column2)");
        winqEqual(new WindowDef().partitionBy(new Column("column1").add(1)),
                "(PARTITION BY column1 + 1)");
        winqEqual(new WindowDef().partitionBy(new ExpressionConvertible[]{
                new Column("column1").add(1), new Column("column2")}),
                "(PARTITION BY column1 + 1, column2)");
        winqEqual(new WindowDef().orderBy(new Column("column1").order(Order.Asc)),
                "(ORDER BY column1 ASC)");
        winqEqual(new WindowDef().frameSpec(new FrameSpec().range().unboundedPreceding()),
                "(RANGE UNBOUNDED PRECEDING)");
        winqEqual(new WindowDef().partitionBy(new ExpressionConvertible[]{
                new Column("column1").add(1), new Column("column2")})
                        .orderBy(new Column("column1").order(Order.Asc))
                        .frameSpec(new FrameSpec().range().unboundedPreceding()),
                "(PARTITION BY column1 + 1, column2 ORDER BY column1 ASC RANGE UNBOUNDED PRECEDING)");
    }
}

File: src/rust/examples/tests/winq/window_def_test.rs
#[cfg(test)]
pub mod window_def_test {
    use crate::base::winq_tool::WinqTool;
    use wcdb::winq::column::{Column, ColumnTrait};
    use wcdb::winq::expression::Expression;
    use wcdb::winq::expression_operable::ExpressionOperableTrait;
    use wcdb::winq::frame_spec::FrameSpec;
    use wcdb::winq::ordering_term::{Order, OrderingTerm};
    use wcdb::winq::window_def::WindowDef;

    #[test]
    pub fn test() {
        WinqTool::winq_equal(&WindowDef::new(), "");

        let window_def = WindowDef::new().partition_by_with_string(&vec!["column1"]);
        WinqTool::winq_equal(&window_def, "(PARTITION BY column1)");

        let window_def = WindowDef::new().partition_by_with_string(&vec!["column1", "column2"]);
        WinqTool::winq_equal(&window_def, "(PARTITION BY column1, column2)");

        let column1 = Column::new("column1", None).add(1);
        let window_def = WindowDef::new().partition(&vec![&column1]);
        WinqTool::winq_equal(&window_def, "(PARTITION BY column1 + 1)");

        let column1 = Column::new("column1", None).add(1);
        let column2 = Column::new("column2", None);
        let expression = Expression::new(&column2);
        let window_def = WindowDef::new().partition(&vec![&column1, &expression]);
        WinqTool::winq_equal(&window_def, "(PARTITION BY column1 + 1, column2)");

        let ordering_term: OrderingTerm = Column::new("column1", None).order(Order::Asc);
        let window_def = WindowDef::new().order_by(&vec![&ordering_term]);
        WinqTool::winq_equal(&window_def, "(ORDER BY column1 ASC)");

        let frame_spec = FrameSpec::new().range().unbounded_preceding();
        let window_def = WindowDef::new().frame_spec(&frame_spec);
        WinqTool::winq_equal(&window_def, "(RANGE UNBOUNDED PRECEDING)");

        let column1 = Column::new("column1", None).add(1);
        column1.add(1);
        let column2 = Column::new("column2", None);
        let expression = Expression::new(&column2);
        let ordering_term: OrderingTerm = Column::new("column1", None).order(Order::Asc);
        let frame_spec = FrameSpec::new().range().unbounded_preceding();
        let window_def = WindowDef::new()
            .partition(&vec![&column1, &expression])
            .order_by(&vec![&ordering_term])
            .frame_spec(&frame_spec);
        WinqTool::winq_equal(
            &window_def,
            "(PARTITION BY column1 + 1, column2 ORDER BY column1 ASC RANGE UNBOUNDED PRECEDING)",
        );
    }
}
